src/rust/src/lib.rs

```diff
@@ -2,7 +2,72 @@
 // 2.0, and the BSD License. See the LICENSE file in the root of this repository
 // for complete details.
 
+/// Returns the value of the input with the most-significant-bit copied to all
+/// of the bits.
+fn duplicate_msb_to_all(a: u16) -> u16 {
+    (1 - (a >> 15)).wrapping_sub(1)
+}
+
+/// This returns 0xFFFF if a < b else 0x0000, but does so in a constant time
+/// fashion.
```

Comment:
Pretty sure that "docstring" is not accurate; here's how I believe it should look. I presume this is a known limitation, but the impact of it on the rest of the code here is not obvious to me.
```suggestion
/// This returns 0xFFFF if a < b (more specifically: 0 < b - a < 2^15 OR a - b >= 2^15),
/// else 0x0000, and does so in a constant-time fashion.
```
If I'm not gravely misunderstanding something, the function should look roughly like this:
```
b /\
  |          2^15        2^16-1
 _|___________|___________|
  |          /           /|
  | 0x0000 /           /  |
  |      /           /    |
  |    /           /      |
  |  /   0xFFFF  /        |
 _|/           /          |
  |          /           /|   
  |        /  0x0000   /  |
  |      /           /    |
  |    /           /      |
  |  /           / 0xFFFF |
 _|/__________ /__________|__\ a
  |           |           |  /
  0          2^15        2^16-1
```

Reply:
Are you saying there's a behvioral difference between the Rust and the C implementations, or that the C implementation also has this issue?

Reply:
The way I understand it, both implementations behave the same.

I was just trying to follow what the implementation was doing... Since the output of `duplicate_msb_to_all` (as well as `Cryptography_DUPLICATE_MSB_TO_ALL`) switches every 2^15 (i.e. whether `a >> 15` is 0 or 1, together with the fact that we wrap at 2^16), and the difference `a - b` can be anything in `[-(2^16-1), 2^16 -1]`, it cannot be constant for all possible combinations of `a < b`.

Reply:
Interesting, the origin of this code appears to have been derived from OpenSSL:

https://github.com/openssl/openssl/commit/5a3d21c0585064292bde5cd34089e120487ab687#diff-a1876514e966a072053ae160d38ab53d245761e72e32281ecb9d16e7a6fe1b0eL70

However, it looks like their modern lt logic is different: https://github.com/openssl/openssl/blob/master/include/internal/constant_time.h#L120

Reply:
Yeah, with those XORs & ORs it sounds more plausible (didn't verify though).

Reply:
Do you have a specific test case that this fails on, that we can add? AFACIT this issue isn't reachable in actual code, since one of the values is always known to be in `[0, 256)`, does that look right to you?

PS: Thanks so much for taking a look at this code!

Reply:
I've gone ahead and swithed this to use `u8` everywhere, which I think fixes the issue, it was a bug that `u16` was ever used in that code IMO -- it reflect a confusion between bits (which are >256) and bytes (which aren't).

Reply:
You're welcome - I was just checking this out of curiosity, and with some mathematical (but very little cryptographic) background. I believe the problem can still occur even if `a` or `b` are restricted to 0-256 (though the problematic area obviously gets much smaller, proportionally).

Still, I'd prefer the variant from openssl and in any case, would recommend not relying on my musings but rather double-checking with the openssl-people. 🙃